itest/lnd_channel_backup_test.go

```diff
@@ -665,6 +782,170 @@ func runChanRestoreScenarioCommitTypes(ht *lntest.HarnessTest,
+		cipherSeed.Entropy[:], dave.Cfg.NetParams,
+	)
+	require.NoError(ht, err)
+
+	// Create keyRing.
+	keyRing := &hdKeyRing{
+		ExtendedKey: extendedRootKey,
+		ChainParams: dave.Cfg.NetParams,
+	}
+
+	// Create signer.
+	signer := &hdSigner{
+		ExtendedKey: extendedRootKey,
+		ChainParams: dave.Cfg.NetParams,
+	}
+	signer.MusigSessionManager = input.NewMusigSessionManager(
+		signer.FetchPrivateKey,
+	)
+
+	// readBackup reads, decrypts and parses backup file. It also returns
+	// raw bytes of channel.backup file.
+	readBackup := func() (chanbackup.Single, []byte) {
+		// Read the entire Multi backup stored within this node's
+		// channels.backup file.
+		multi, err := os.ReadFile(backupFilePath)
+		require.NoError(ht, err)
+
+		// Decrypt and parse the file.
+		var m chanbackup.Multi
+		err = m.UnpackFromReader(bytes.NewReader(multi), keyRing)
+		require.NoError(ht, err)
+		require.Equal(ht, 1, len(m.StaticBackups))
+		single := m.StaticBackups[0]
+		require.NotNil(ht, single.CloseTxInputs)
+		require.NotNil(ht, single.CloseTxInputs.CommitTx)
+		require.NotEmpty(ht, single.CloseTxInputs.CommitSig)
+
+		return single, multi
+	}
+
+	// outputsAsString formats values of transaction outputs as a comma
+	// separated string.
+	outputsAsString := func(tx *wire.MsgTx) string {
+		outputs := make([]string, 0, len(tx.TxOut))
+		for _, txOut := range tx.TxOut {
+			value := txOut.Value
+			outputs = append(outputs, fmt.Sprintf("%d", value))
+		}
+
+		return strings.Join(outputs, ",")
+	}
+
+	// Make sure the nodes are connected for the payment to pass.
+	ht.EnsureConnected(carol, dave)
+
+	// Send some sats from Dave to Carol. The channel state will change,
+	// but not in the backup.
+	payReqs, _, _ := ht.CreatePayReqs(carol, 100_000, 1)
+	ht.CompletePaymentRequests(dave, payReqs)
+
+	// Snapshot channel backup state before Dave's shutdown.
+	backupBeforeShutdown, _ := readBackup()
+	balancesBeforeShutdown := outputsAsString(
+		backupBeforeShutdown.CloseTxInputs.CommitTx,
+	)
+
+	// Now shutdown Dave. It should update the backup file upon shutdown.
+	// Use method SuspendNode() instead of Shutdown() to prevent node's
+	// directory from being cleaned up, so we can read channel.backup.
+	ht.SuspendNode(dave)
+
+	// Snapshot channel backup state after Dave's shutdown.
+	backupAfterShutdown, multiAfterShutdown := readBackup()
+	balancesAfterShutdown := outputsAsString(
+		backupAfterShutdown.CloseTxInputs.CommitTx,
+	)
+
+	// Make sure balances in the backup changed during shutdown.
+	require.NotEqual(ht, balancesBeforeShutdown, balancesAfterShutdown)
+
+	// Generate force close transaction from SCB to use it to restore.
+	var scbForCloseTx chanbackup.Single
+	if useOutdatedScbToClose {
+		// Intentionally use an outdated SCB to test Carol's penalty.
+		scbForCloseTx = backupBeforeShutdown
+	} else {
+		// Use the latest SCB to produce force close tx.
+		scbForCloseTx = backupAfterShutdown
+	}
+	crs.forceCloseTx, err = chanbackup.SignCloseTx(
```

Comment:
Instead of creating key rings and signers outside of the `SignCloseTx`, I think it should be handled inside the method. This way we only need to pass `scbForCloseTx` and `mnemonic` and `password`, and the `chanbackup` handles the creation of signers and king rings.

Also by looking at this method alone it's not an RPC point? If it's a method meant to be imported by other packages to be used, I'm not sure if it should be tested in integration tests.

Reply:
If we make it too easy for the user to make signed transactions, they are more likely to misuse the feature and get funds lost. Currently it is not trivial to use the function outside of chantools, where we warn the user many times. The implementations of dependencies (keychain.KeyRing, keychain.ECDHRing, and input.Signer) live as private types in itest and are not directly usable from outside.

This function is intended to be used from chantools only. We don't have chantools in itest, so the only way of emulating using chantools in itest is to call the function directly. The code in chantools is really a wrapper around this function: https://github.com/lightninglabs/chantools/pull/95/commits/1fe8248114f0c6f453ab28b99b3d74abdcf1943b

Reply:
In that case we should add unit tests for this method, not itest. You can import the `lntest` package in `chantools` to build the exact same test, and we should not use the itest in `lnd` to test programs importing it.

Reply:
@yyforyongyu I removed most itests (using hdsigner and hdkeyring). They were replaced with unit tests in chantools. Now the only itest change left is the check that LND updates channel.backup file upon shutdown, which does not add any new dependencies.